File: mac-app/Sources/AIBotBridge/QuotaService.swift
import Foundation

actor MacQuotaService {
    private let store: MacDataStore
    private let session: URLSession

    init(store: MacDataStore) {
        self.store = store
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 15
        configuration.timeoutIntervalForResource = 20
        session = URLSession(configuration: configuration)
    }

    func refresh() async {
        async let claude = fetchClaude()
        async let codex = fetchCodex()
        let values = await (claude, codex)
        store.mergeQuotas(claude: values.0, codex: values.1)
    }

    static func parseClaude(_ data: Data, credentialPlan: String? = nil,
                            now: Date = Date()) throws -> ProviderQuotaSnapshot {
        let root = try object(data)
        let primary = dictionary(root["five_hour"])
        let weekly = dictionary(root["seven_day"])
        return ProviderQuotaSnapshot(
            provider: "claude",
            plan: text(root["plan_type"]) ?? text(root["subscription_type"]) ?? credentialPlan,
            primaryPercent: number(primary?["utilization"]),
            primaryResetsAt: date(primary?["resets_at"]),
            weeklyPercent: number(weekly?["utilization"]),
            weeklyResetsAt: date(weekly?["resets_at"]),
            resetCreditsAvailable: nil,
            resetCreditExpiresAt: [],
            updatedAt: now,
            stale: false)
    }

    static func parseCodex(_ usage: Data, credits: Data? = nil, credentialPlan: String? = nil,
                           now: Date = Date()) throws -> ProviderQuotaSnapshot {
        let root = try object(usage)
        guard let rateLimit = dictionary(root["rate_limit"]) else { throw QuotaError.invalidResponse }
        var primaryPercent: Double?
        var primaryReset: Date?
        var weeklyPercent: Double?
        var weeklyReset: Date?
        for key in ["primary_window", "secondary_window"] {
            guard let window = dictionary(rateLimit[key]) else { continue }
            if (number(window["limit_window_seconds"]) ?? 0) >= 172_800 {
                weeklyPercent = number(window["used_percent"])
                weeklyReset = unixDate(window["reset_at"])
            } else {
                primaryPercent = number(window["used_percent"])
                primaryReset = unixDate(window["reset_at"])
            }
        }

        var available = integer(dictionary(root["rate_limit_reset_credits"])?["available_count"])
        var expirations: [Int64] = []
        if let credits {
            let creditRoot = try object(credits)
            available = integer(creditRoot["available_count"]) ?? available
            for row in creditRoot["credits"] as? [[String: Any]] ?? []
                where text(row["status"])?.lowercased() == "available" {
                if let expiration = date(row["expires_at"]) {
                    expirations.append(Int64(expiration.timeIntervalSince1970))
                }
            }
            expirations.sort()
        }
        return ProviderQuotaSnapshot(
            provider: "codex",
            plan: text(root["plan_type"]) ?? credentialPlan,
            primaryPercent: primaryPercent,
            primaryResetsAt: primaryReset,
            weeklyPercent: weeklyPercent,
            weeklyResetsAt: weeklyReset,
            resetCreditsAvailable: available,
            resetCreditExpiresAt: expirations,
            updatedAt: now,
            stale: false)
    }

    private func fetchClaude() async -> ProviderQuotaSnapshot? {
        guard let credential = Self.readClaudeCredential() else { return nil }
        var request = URLRequest(url: URL(string: "https://api.anthropic.com/api/oauth/usage")!)
        request.setValue("Bearer \(credential.token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("oauth-2025-04-20", forHTTPHeaderField: "anthropic-beta")
        request.setValue("claude-code/2.1.0", forHTTPHeaderField: "User-Agent")
        do {
            let (data, response) = try await session.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
            return try Self.parseClaude(data, credentialPlan: credential.plan)
        } catch {
            return nil
        }
    }

    private func fetchCodex() async -> ProviderQuotaSnapshot? {
        guard let credential = Self.readCodexCredential() else { return nil }
        do {
            guard let usage = try await sendCodex(
                "https://chatgpt.com/backend-api/wham/usage", credential: credential) else { return nil }
            let credits = try await sendCodex(
                "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits", credential: credential)
            return try Self.parseCodex(usage, credits: credits, credentialPlan: credential.plan)
        } catch {
            return nil
        }
    }

    private func sendCodex(_ address: String, credential: CodexCredential) async throws -> Data? {
        var request = URLRequest(url: URL(string: address)!)
        request.setValue("Bearer \(credential.token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("AI-bot/0.1", forHTTPHeaderField: "User-Agent")
        if let accountID = credential.accountID {
            request.setValue(accountID, forHTTPHeaderField: "ChatGPT-Account-Id")
        }
        let (data, response) = try await session.data(for: request)
        return (response as? HTTPURLResponse)?.statusCode == 200 ? data : nil
    }

    private static func readClaudeCredential() -> (token: String, plan: String?)? {
        let path = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".claude/.credentials.json")
        guard let data = try? Data(contentsOf: path),
              let root = try? object(data),
              let oauth = dictionary(root["claudeAiOauth"]),
              let token = text(oauth["accessToken"]),
              !token.isEmpty else { return nil }
        return (token, text(oauth["subscriptionType"]) ?? text(oauth["subscription_type"]))
    }

    private static func readCodexCredential() -> CodexCredential? {
        let path = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex/auth.json")
        guard let data = try? Data(contentsOf: path),
              let root = try? object(data),
              let tokens = dictionary(root["tokens"]),
              let token = text(tokens["access_token"]),
              !token.isEmpty,
              !jwtExpired(token) else { return nil }
        var accountID = text(tokens["account_id"])
        var plan: String?
        if let idToken = text(tokens["id_token"]),
           let claims = jwtClaims(idToken),
           let auth = dictionary(claims["https://api.openai.com/auth"]) {
            accountID = accountID ?? text(auth["chatgpt_account_id"])
            plan = text(auth["chatgpt_plan_type"])
        }
        return CodexCredential(token: token, accountID: accountID, plan: plan)
    }

    private static func jwtExpired(_ token: String) -> Bool {
        guard let claims = jwtClaims(token), let expiration = number(claims["exp"]) else { return false }
        return expiration <= Date().timeIntervalSince1970 + 60
    }

    private static func jwtClaims(_ token: String) -> [String: Any]? {
        let parts = token.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1]).replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        payload += String(repeating: "=", count: (4 - payload.count % 4) % 4)
        guard let data = Data(base64Encoded: payload) else { return nil }
        return try? object(data)
    }

    private static func object(_ data: Data) throws -> [String: Any] {
        guard let value = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw QuotaError.invalidResponse
        }
        return value
    }

    private static func dictionary(_ value: Any?) -> [String: Any]? { value as? [String: Any] }
    private static func text(_ value: Any?) -> String? { value as? String }

    private static func number(_ value: Any?) -> Double? {
        if let number = value as? NSNumber { return number.doubleValue }
        if let text = value as? String { return Double(text) }
        return nil
    }

    private static func integer(_ value: Any?) -> Int? {
        guard let number = number(value) else { return nil }
        return Int(number)
    }

    private static func date(_ value: Any?) -> Date? {
        guard let text = text(value) else { return nil }
        let formatter = ISO8601DateFormatter()
        if let parsed = formatter.date(from: text) { return parsed }
        formatter.formatOptions.insert(.withFractionalSeconds)
        return formatter.date(from: text)
    }

    private static func unixDate(_ value: Any?) -> Date? {
        guard let seconds = number(value) else { return nil }
        return Date(timeIntervalSince1970: seconds)
    }

    private struct CodexCredential {
        let token: String
        let accountID: String?
        let plan: String?
    }

    private enum QuotaError: Error { case invalidResponse }
}

File: windows-app/AIBotBridge/QuotaService.cs
using System.Text;
using System.Text.Json;

namespace AIBotBridge;

internal sealed class QuotaService
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(15) };
    private readonly object _sync = new();
    private QuotaSnapshot? _snapshot;

    internal QuotaService()
    {
        var cached = SnapshotCache.Load<QuotaSnapshot>("usage-cache.json");
        if (cached is not null)
            _snapshot = new QuotaSnapshot(Stale(cached.Claude), Stale(cached.Codex));
    }

    internal QuotaSnapshot? Snapshot
    {
        get { lock (_sync) return _snapshot; }
    }

    internal async Task RunAsync(CancellationToken cancellationToken)
    {
        await RefreshAsync(cancellationToken);
        using var timer = new PeriodicTimer(TimeSpan.FromMinutes(2));
        while (await timer.WaitForNextTickAsync(cancellationToken))
            await RefreshAsync(cancellationToken);
    }

    internal async Task RefreshAsync(CancellationToken cancellationToken)
    {
        var claudeTask = FetchClaudeAsync(cancellationToken);
        var codexTask = FetchCodexAsync(cancellationToken);
        await Task.WhenAll(claudeTask, codexTask);

        lock (_sync)
        {
            var claude = claudeTask.Result ?? Stale(_snapshot?.Claude);
            var codex = MergeCodex(codexTask.Result, _snapshot?.Codex);
            if (claude is null && codex is null) return;
            _snapshot = new QuotaSnapshot(claude, codex);
            if (claude?.Stale == false || codex?.Stale == false)
                SnapshotCache.Save("usage-cache.json", _snapshot);
        }
    }

    internal static ProviderQuotaSnapshot? MergeCodex(ProviderQuotaSnapshot? fresh, ProviderQuotaSnapshot? previous)
    {
        if (fresh is null) return Stale(previous);
        if (fresh.ResetCreditsAvailable is > 0 && fresh.ResetCreditsAvailable == previous?.ResetCreditsAvailable &&
            fresh.ResetCreditExpiresAt.Count == 0 && previous.ResetCreditExpiresAt.Count > 0)
            return fresh with { ResetCreditExpiresAt = previous.ResetCreditExpiresAt.ToArray(), Stale = true };
        return fresh;
    }

    internal static ProviderQuotaSnapshot ParseClaude(string json, string? credentialPlan = null)
    {
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;
        var fiveHour = root.TryGetProperty("five_hour", out var primary) ? primary : default;
        var sevenDay = root.TryGetProperty("seven_day", out var weekly) ? weekly : default;
        return new ProviderQuotaSnapshot(
            "claude",
            String(root, "plan_type") ?? String(root, "subscription_type") ?? credentialPlan,
            Number(fiveHour, "utilization"), Date(fiveHour, "resets_at"),
            Number(sevenDay, "utilization"), Date(sevenDay, "resets_at"),
            null, Array.Empty<long>(), DateTimeOffset.UtcNow, false);
    }

    internal static ProviderQuotaSnapshot ParseCodex(string usageJson, string? creditsJson = null,
        string? credentialPlan = null)
    {
        using var document = JsonDocument.Parse(usageJson);
        var root = document.RootElement;
        if (!root.TryGetProperty("rate_limit", out var rateLimit))
            throw new JsonException("rate_limit is missing.");

        double? primaryPercent = null, weeklyPercent = null;
        DateTimeOffset? primaryReset = null, weeklyReset = null;
        foreach (var name in new[] { "primary_window", "secondary_window" })
        {
            if (!rateLimit.TryGetProperty(name, out var window) || window.ValueKind != JsonValueKind.Object)
                continue;
            var seconds = Number(window, "limit_window_seconds");
            var isWeekly = seconds is >= 172800;
            if (isWeekly)
            {
                weeklyPercent = Number(window, "used_percent");
                weeklyReset = UnixDate(window, "reset_at");
            }
            else
            {
                primaryPercent = Number(window, "used_percent");
                primaryReset = UnixDate(window, "reset_at");
            }
        }

        int? available = null;
        var expirations = new List<long>();
        if (root.TryGetProperty("rate_limit_reset_credits", out var embedded))
            available = Integer(embedded, "available_count");
        if (!string.IsNullOrWhiteSpace(creditsJson))
        {
            using var creditsDocument = JsonDocument.Parse(creditsJson);
            var credits = creditsDocument.RootElement;
            available = Integer(credits, "available_count") ?? available;
            if (credits.TryGetProperty("credits", out var rows) && rows.ValueKind == JsonValueKind.Array)
            {
                foreach (var row in rows.EnumerateArray())
                {
                    if (!string.Equals(String(row, "status"), "available", StringComparison.OrdinalIgnoreCase))
                        continue;
                    var expires = Date(row, "expires_at");
                    if (expires.HasValue) expirations.Add(expires.Value.ToUnixTimeSeconds());
                }
                expirations.Sort();
            }
        }

        return new ProviderQuotaSnapshot(
            "codex", String(root, "plan_type") ?? credentialPlan,
            primaryPercent, primaryReset, weeklyPercent, weeklyReset,
            available, expirations, DateTimeOffset.UtcNow, false);
    }

    private static async Task<ProviderQuotaSnapshot?> FetchClaudeAsync(CancellationToken cancellationToken)
    {
        var credential = ReadClaudeCredential();
        if (credential is null) return null;
        using var request = new HttpRequestMessage(HttpMethod.Get, "https://api.anthropic.com/api/oauth/usage");
        request.Headers.TryAddWithoutValidation("Authorization", "Bearer " + credential.Value.Token);
        request.Headers.TryAddWithoutValidation("Accept", "application/json");
        request.Headers.TryAddWithoutValidation("anthropic-beta", "oauth-2025-04-20");
        request.Headers.TryAddWithoutValidation("User-Agent", "claude-code/2.1.0");
        try
        {
            using var response = await Http.SendAsync(request, cancellationToken);
            if (!response.IsSuccessStatusCode) return null;
            return ParseClaude(await response.Content.ReadAsStringAsync(cancellationToken), credential.Value.Plan);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or JsonException)
        {
            return null;
        }
    }

    private static async Task<ProviderQuotaSnapshot?> FetchCodexAsync(CancellationToken cancellationToken)
    {
        var credential = ReadCodexCredential();
        if (credential is null) return null;
        try
        {
            var usage = await SendCodexAsync(
                "https://chatgpt.com/backend-api/wham/usage", credential.Value, cancellationToken);
            if (usage is null) return null;
            var credits = await SendCodexAsync(
                "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits",
                credential.Value, cancellationToken);
            return ParseCodex(usage, credits, credential.Value.Plan);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or JsonException)
        {
            return null;
        }
    }

    private static async Task<string?> SendCodexAsync(
        string url, (string Token, string? AccountId, string? Plan) credential,
        CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.TryAddWithoutValidation("Authorization", "Bearer " + credential.Token);
        request.Headers.TryAddWithoutValidation("Accept", "application/json");
        request.Headers.TryAddWithoutValidation("User-Agent", "AI-bot/0.1");
        if (!string.IsNullOrWhiteSpace(credential.AccountId))
            request.Headers.TryAddWithoutValidation("ChatGPT-Account-Id", credential.AccountId);
        using var response = await Http.SendAsync(request, cancellationToken);
        return response.IsSuccessStatusCode
            ? await response.Content.ReadAsStringAsync(cancellationToken) : null;
    }

    private static (string Token, string? Plan)? ReadClaudeCredential()
    {
        var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".claude", ".credentials.json");
        try
        {
            using var document = JsonDocument.Parse(File.ReadAllText(path));
            var oauth = document.RootElement.GetProperty("claudeAiOauth");
            var token = String(oauth, "accessToken");
            return string.IsNullOrWhiteSpace(token) ? null
                : (token, String(oauth, "subscriptionType") ?? String(oauth, "subscription_type"));
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or JsonException or KeyNotFoundException)
        {
            return null;
        }
    }

    private static (string Token, string? AccountId, string? Plan)? ReadCodexCredential()
    {
        var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".codex", "auth.json");
        try
        {
            using var document = JsonDocument.Parse(File.ReadAllText(path));
            var tokens = document.RootElement.GetProperty("tokens");
            var access = String(tokens, "access_token");
            if (string.IsNullOrWhiteSpace(access) || JwtExpired(access)) return null;
            var accountId = String(tokens, "account_id");
            string? plan = null;
            var idToken = String(tokens, "id_token");
            if (!string.IsNullOrWhiteSpace(idToken) && TryReadJwt(idToken, out var claims))
            {
                if (claims.TryGetProperty("https://api.openai.com/auth", out var auth))
                {
                    accountId ??= String(auth, "chatgpt_account_id");
                    plan = String(auth, "chatgpt_plan_type");
                }
            }
            return (access, accountId, plan);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or JsonException or KeyNotFoundException)
        {
            return null;
        }
    }

    private static bool JwtExpired(string token)
    {
        if (!TryReadJwt(token, out var claims)) return false;
        var expires = Number(claims, "exp");
        return expires.HasValue && expires.Value <= DateTimeOffset.UtcNow.ToUnixTimeSeconds() + 60;
    }

    private static bool TryReadJwt(string token, out JsonElement claims)
    {
        claims = default;
        try
        {
            var parts = token.Split('.');
            if (parts.Length < 2) return false;
            var payload = parts[1].Replace('-', '+').Replace('_', '/');
            payload = payload.PadRight(payload.Length + (4 - payload.Length % 4) % 4, '=');
            using var document = JsonDocument.Parse(Encoding.UTF8.GetString(Convert.FromBase64String(payload)));
            claims = document.RootElement.Clone();
            return true;
        }
        catch (Exception ex) when (ex is FormatException or JsonException)
        {
            return false;
        }
    }

    private static ProviderQuotaSnapshot? Stale(ProviderQuotaSnapshot? value) =>
        value is null ? null : value with { Stale = true };

    private static string? String(JsonElement parent, string name) =>
        parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out var value) &&
        value.ValueKind == JsonValueKind.String ? value.GetString() : null;

    private static double? Number(JsonElement parent, string name) =>
        parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out var value) &&
        value.ValueKind == JsonValueKind.Number ? value.GetDouble() : null;

    private static int? Integer(JsonElement parent, string name) =>
        parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out var value) &&
        value.TryGetInt32(out var number) ? number : null;

    private static DateTimeOffset? Date(JsonElement parent, string name) =>
        DateTimeOffset.TryParse(String(parent, name), out var value) ? value : null;

    private static DateTimeOffset? UnixDate(JsonElement parent, string name)
    {
        var value = Number(parent, name);
        return value.HasValue ? DateTimeOffset.FromUnixTimeSeconds((long)value.Value) : null;
    }
}
